asm/exceptasm.s

```diff
@@ -69,646 +81,893 @@ glabel __osIntTable
+    li      $at, ~CAUSE_IP4
+    and     $s0, $s0, $at
+    # If the callback is NULL, handling is done
+    beqz    $t2, send_cart_mesg
+     addi   $t1, $t1, (OS_INTR_CART*8)
+    # Set up a stack and run the callback
+    jalr    $t2
+    lw      $sp, HWINT_SP($t1)
+    beqz    $v0, send_cart_mesg
+     nop
+    # Redispatch immediately if the callback returned nonzero
+    b       redispatch
+     nop
+send_cart_mesg:
+    # Post a cart event message
+    jal     send_mesg
+     li     $a0, OS_EVENT_CART*8
+    # Continue
+    b       next_interrupt
+     nop
+
+/**
+ *  IP3/RCP Interrupt
+ *  Signalled by the RCP for various reasons, described below
+ */
 rcp:
-/* 004894 80003C94 3C088001 */  lui   $t0, %hi(__OSGlobalIntMask) # $t0, 0x8001
-/* 004898 80003C98 2508AD00 */  addiu $t0, %lo(__OSGlobalIntMask) # addiu $t0, $t0, -0x5300
-/* 00489C 80003C9C 8D080000 */  lw    $t0, ($t0)
-/* 0048A0 80003CA0 3C11A430 */  lui   $s1, %hi(D_A4300008) # $s1, 0xa430
-/* 0048A4 80003CA4 8E310008 */  lw    $s1, %lo(D_A4300008)($s1)
-/* 0048A8 80003CA8 00084402 */  srl   $t0, $t0, 0x10
-/* 0048AC 80003CAC 02288824 */  and   $s1, $s1, $t0
-/* 0048B0 80003CB0 32290001 */  andi  $t1, $s1, 1
-/* 0048B4 80003CB4 11200013 */  beqz  $t1, vi
-/* 0048B8 80003CB8 00000000 */   nop   
-/* 0048BC 80003CBC 3C0CA404 */  lui   $t4, %hi(D_A4040010) # $t4, 0xa404
-/* 0048C0 80003CC0 8D8C0010 */  lw    $t4, %lo(D_A4040010)($t4)
-/* 0048C4 80003CC4 34098008 */  li    $t1, 32776
-/* 0048C8 80003CC8 3C01A404 */  lui   $at, %hi(D_A4040010) # $at, 0xa404
-/* 0048CC 80003CCC 318C0300 */  andi  $t4, $t4, 0x300
-/* 0048D0 80003CD0 3231003E */  andi  $s1, $s1, 0x3e
-/* 0048D4 80003CD4 11800007 */  beqz  $t4, sp_other_break
-/* 0048D8 80003CD8 AC290010 */   sw    $t1, %lo(D_A4040010)($at)
-/* 0048DC 80003CDC 0C000FCB */  jal   send_mesg
-/* 0048E0 80003CE0 24040020 */   li    $a0, 32
-/* 0048E4 80003CE4 12200042 */  beqz  $s1, NoMoreRcpInts
-/* 0048E8 80003CE8 00000000 */   nop   
-/* 0048EC 80003CEC 10000005 */  b     vi
-/* 0048F0 80003CF0 00000000 */   nop   
+    # Load the MI interrupts and mask with the global interrupt mask
+    lui     $t0, %hi(__OSGlobalIntMask)
+    addiu   $t0, %lo(__OSGlobalIntMask)
+    lw      $t0, ($t0)
+    lui     $s1, %hi(PHYS_TO_K1(MI_INTR_REG))
+    lw      $s1, %lo(PHYS_TO_K1(MI_INTR_REG))($s1)
+    srl     $t0, $t0, 0x10
+    and     $s1, $s1, $t0
+
+/**
+ *  Signal Processor (SP) Interrupt
+ */
+sp:
+    # Test for sp interrupt
+    andi    $t1, $s1, MI_INTR_SP
+    beqz    $t1, vi
+     nop
+    # Test for yielded or done signals in particular
+    lui     $t4, %hi(PHYS_TO_K1(SP_STATUS_REG))
+    lw      $t4, %lo(PHYS_TO_K1(SP_STATUS_REG))($t4)
+    li      $t1, (SP_CLR_INTR | SP_CLR_SIG3)
+    lui     $at, %hi(PHYS_TO_K1(SP_STATUS_REG))
+    andi    $t4, $t4, (SP_STATUS_YIELDED | SP_STATUS_TASKDONE)
+    # Mask out SP interrupt
+    andi    $s1, $s1, (MI_INTR_MASK_SI | MI_INTR_MASK_AI | MI_INTR_MASK_VI | MI_INTR_MASK_PI | MI_INTR_MASK_DP)
```

Comment:
since s1 comes from `MI_INTR_REG` and not `MI_INTR_MASK_REG` shouldn't it use the `MI_INTR_*` defines?
```suggestion
    andi    $s1, $s1, (MI_INTR_SI | MI_INTR_AI | MI_INTR_VI | MI_INTR_PI | MI_INTR_DP)
```